### ISSUE TITLE
Fitness function-specific arguments

### ISSUE THREAD

**Issue by jmmcd** [jmmcd/PonyGE2#24]
I have a fitness function which is really a family of functions, i.e. it is parameterised, so the natural way to write it is along these lines:

```python
class MyFitness:
    def __init__(self, a, b, c):
        self.a = a
        # etc
    def __call__(self, x):
        # use self.a to control fitness evaluation
```

I think we discussed how to pass parameters like `a` in on the command line, without coming to a conclusion. One option would be a new parameter named, say, `FITNESS_PARAMETERS`, passed in a string, and then `eval`-ed or whatever inside the fitness function. That would lead to this type of code:

```python
class MyFitness:
    def __init__(self):
        self.a, self.b, self.c = eval(parameters['FITNESS_PARAMETERS'])
        # etc
```
Then one would write on the command line:
```bash
$ ponyge.py --FITNESS_PARAMETERS "(3, 4, 5)"
```

I've implemented this and it works for me... does this make sense or does anyone have anything better?

**Comment by mikefenton** [comment]
That makes sense, the other option that I have been doing with the bell networks stuff is to simply specify any extra parameters in a params file; they get loaded in and set automatically, so you can add whatever you want there.

**Comment by dvpfagan** [comment]
Why not both?

**Comment by mikefenton** [comment]
One involves changing some ponyge code by adding new if/else clauses in parameters.py, the other bypasses the main code altogether but is susceptible to spelling errors (leading to unintended parameters being added to the params dict, such as "CROSOVER" or what have you).

**Comment by t-h-e** [comment]
I think I mentioned it before: [YAML](https://en.wikipedia.org/wiki/YAML)
Would be an easy way to have the parameters in a more structured form. Arguments for different fitness functions can be put in different sections.

Maybe every fitness function should have a common base class, which provides a method to check if all required parameters are set, because spelling errors and missing parameters are possible with every method mentioned so far.

**Comment by dvpfagan** [comment]
Having a generic parameter called fitness_params makes sense and then the
fitness function can access it if needs be. Only requires a single if else
in ponyge

we could look into using something like this

def someFunc(arg1, arg2, arg3, *args, **kwargs)
    for x in args
        print x

This sort of thing would allow any amount of params to be passed to the
fitness function but for me adding the fitness_params thing to the
params dictionary would accomplish the same thing. We just document it as a
variable sized input to be used with a fitness function. So long as we
don;t eval it until the fitness function it shouldn't mess up stuff like
crossover etc as it would define new variable called crossover or whatever
in the scope of the fitness function only.

Dave

**Comment by mfjoneill** [comment]
Hi all, I'd vote for Dave's combination approach to have both available. I
find it useful to set some named parameters from the command line (i.e. the
ones under experimental analysis) and from shell scripts while using the
parameters file to set default settings for all other parameters. The new
approach wouldn't be my cup of tea as I'd loose track of which numbers
belonged to which parameter ;-)
Many thanks,
Mike.

**Comment by mikefenton** [comment]
The plan was never to get rid of either option, but to present two options to the user for creating new problems. They can either specify new parameters directly in the params file, where they will be read in and loaded into the params dictionary automatically (currently an already existing solution, i.e. we don't have to change the code for this), or they can give some command line input which will be matched to an if/else clause in parameters.set_params(). 

For the latter command-line option we simply need to figure out the best way of adding in new parameters from the command line, as James mentioned in the original post at the top of this thread. I think James's solution for variable length/multiple inputs is pretty elegant, but are there any other suggestions for achieving the same thing?

**Comment by jmmcd** [comment]
We have decided to name the new parameter `EXTRA_FITNESS_PARAMETERS`, which is a string to be interpreted and used by the fitness object in any way it wants.

It's called `EXTRA` because we are retaining `ERROR_METRIC`, `DATASET`, and `TARGET`. These seem quite generic parameters which could be used across many fitness functions.

We haven't changed the default dictionary + params.txt file + command-line mechanism for now, but we talked about it!
